### PR TITLE
feat: support new `otlp` receiver configuration

### LINKED ISSUE #6778: Support new `otlp` receiver configuration
### Component(s)

Prometheus, PrometheusAgent

### What is missing? Please describe.

https://github.com/prometheus/prometheus/pull/14200 allows to configure resource attribute promotion for the OTLP receiver endpoint.
The Prometheus and PrometheusAgent CRDs should be extended to support the same.

The feature should be available with the upcoming v2.54.0 Prometheus.

### Describe alternatives you've considered.

NA

## Environment
Kubernetes Version:
Prometheus-Operator Version:

### PR DESCRIPTION
**Pull request by dongjiang1989**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

Closes #6778 

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
 support new `otlp` receiver configuration
```

### CODE REVIEW

File: pkg/prometheus/promcfg.go [hunk 1]
```diff
@@ -4104,6 +4114,24 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 	return cfg, nil
 }
 
+func (cg *ConfigGenerator) generateOTLPConfig(_ *assets.StoreBuilder) yaml.MapItem {
```
Comment (nicolastakashi) [inline pkg/prometheus/promcfg.go:4117]: wdyt about receive the otlp config as a parameters since on the place where we're invoking generateOTLPConfig we're already checking if the OTLP field from the Common Fields is != nil.

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 2]
```diff
@@ -1998,3 +2004,11 @@ type ScrapeClass struct {
 	// +optional
 	AttachMetadata *AttachMetadata `json:"attachMetadata,omitempty"`
 }
+
+// OTLPConfig is the configuration for writing to the OTLP endpoint.
+//
+// +k8s:openapi-gen=true
+type OTLPConfig struct {
+	// +optional
```
Comment (mviswanathsai) [inline pkg/apis/monitoring/v1/prometheus_types.go:2012]: We could add a description here and also seems reasonable to add a Minimum length validation.
```suggestion
        // List of OTel Attributes that should be promoted to metric labels, defaults to none.
        // +kubebuilder:validation:MinLength=1
	// +optional
```

Comment (mviswanathsai) [reply pkg/apis/monitoring/v1/prometheus_types.go:2012]: We may also want to prevent duplicate values. WDYT?
```suggestion
        // List of OTel Attributes that should be promoted to metric labels, defaults to none.
        // +kubebuilder:validation:MinLength=1
        // +kubebuilder:validation:listType=map
        //
	// +optional
```

File: pkg/prometheus/promcfg.go [hunk 3]
```diff
@@ -770,6 +770,11 @@ func (cg *ConfigGenerator) GenerateServerConfiguration(
 		cfg = append(cfg, cg.generateRemoteReadConfig(remoteRead, store))
 	}
 
+	// OTLP config
+	if cpf.OTLP != nil && cg.version.GTE(semver.MustParse("2.54.0")) {
```
Comment (mviswanathsai) [inline pkg/prometheus/promcfg.go:774]: Should we do `cg.WithMinimumVersion(...)`?
Or we could also reject the object at `resource_selector.go`. 
WDYT? 
CC: @simonpasquier

File: pkg/prometheus/promcfg.go [hunk 4]
```diff
@@ -2540,6 +2545,11 @@ func (cg *ConfigGenerator) GenerateAgentConfiguration(
 		cfg = append(cfg, cg.generateRemoteWriteConfig(store))
 	}
 
+	// OTLP config
+	if cpf.OTLP != nil && cg.version.GTE(semver.MustParse("2.54.0")) {
```
Comment (mviswanathsai) [inline pkg/prometheus/promcfg.go:2549]: Same remark as with the Server code.

File: pkg/prometheus/promcfg.go [hunk 5]
```diff
@@ -4104,6 +4114,24 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 	return cfg, nil
 }
 
+func (cg *ConfigGenerator) generateOTLPConfig(_ *assets.StoreBuilder) yaml.MapItem {
+	cfg := yaml.MapSlice{}
+
+	otlpConfig := cg.prom.GetCommonPrometheusFields().OTLP
+
+	if len(otlpConfig.PromoteResourceAttributes) > 0 {
+		cfg = append(cfg, yaml.MapItem{
```
Comment (mviswanathsai) [inline pkg/prometheus/promcfg.go:4123]: I don't think we need a MapSlice here. This would do the trick:
```suggestion
		cfg := yaml.MapItem{
```

File: pkg/prometheus/promcfg.go [hunk 6]
```diff
@@ -4104,6 +4114,24 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+			Key:   "promote_resource_attributes",
+			Value: otlpConfig.PromoteResourceAttributes,
+		})
+	}
+
+	return yaml.MapItem{
+		Key:   "otlp",
+		Value: cfg,
```
Comment (mviswanathsai) [inline pkg/prometheus/promcfg.go:4131]: We could even go a step further and do the following inside an `if` block.

```suggestion
		Value: yaml.MapItem{
		    Key:   "promote_resource_attributes",
		    Value: otlpConfig.PromoteResourceAttributes,
		    },
```

**Review by mviswanathsai — COMMENTED** [review]
Could you also add an e2e test for an empty `PromoteResourceAttributes` field, and array with duplicate attribute values to make sure the API validations are working as expected?

**Comment by dongjiang1989** [comment]
> Could you also add an e2e test for an empty `PromoteResourceAttributes` field, and array with duplicate attribute values to make sure the API validations are working as expected?

This feature will be available in `2.54.x` Prometheus and it is not released yet. 
So we would need to wait for this till then to add an e2e test

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 2 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/prometheus_types.go:2012]: > We may also want to prevent duplicate values. WDYT?

Great. Verify duplicate values

**Comment by mviswanathsai** [comment]
> > Could you also add an e2e test for an empty `PromoteResourceAttributes` field, and array with duplicate attribute values to make sure the API validations are working as expected?
> 
> This feature will be available in `2.54.x` Prometheus and it is not released yet. 
> So we would need to wait for this till then to add an e2e test 

I see what you mean, what i meant to say was:
1. Run an instance of the operator.
2. Try to create the Prometheus/PrometheusAgent CRDs with the different values for `PromoteResourceAttributes` and see that it is rejected/accepted by the Operator.

IIUC, this does not involve Prometheus itself. I meant to just validate our APIs.

File: pkg/prometheus/promcfg.go (hunk 5 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go:4123]: Thanks.
To be reserved for other fields for `otlpConfig`

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 7]
```diff
@@ -1998,3 +2004,15 @@ type ScrapeClass struct {
 	// +optional
 	AttachMetadata *AttachMetadata `json:"attachMetadata,omitempty"`
 }
+
+// OTLPConfig is the configuration for writing to the OTLP endpoint.
+//
+// +k8s:openapi-gen=true
+type OTLPConfig struct {
+	// List of OTel Attributes that should be promoted to metric labels, defaults to none.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:2012]:
```suggestion
	// List of OpenTelemetry Attributes that should be promoted to metric labels, defaults to none.
```

File: pkg/prometheus/promcfg.go (hunk 1 above)
Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:4117]: we don't even need an otlp config parameter since it's coming from the Prometheus spec for which the config generator has a copy.

File: pkg/prometheus/promcfg.go (hunk 3 above)
Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:774]: In this case, I'd be in favor of failing the reconciliation indeed if the Prometheus version doesn't match the requirement. One idea would be to validate the requirements in NewConfigGenerator and return an error if necessary.

Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go]: Thanks @mviswanathsai @simonpasquier .  If the Prometheus version doesn't match the requirement, return error.

File: pkg/prometheus/promcfg.go [hunk 8]
```diff
@@ -4124,6 +4146,27 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 	return cfg, nil
 }
 
+func (cg *ConfigGenerator) generateOTLPConfig(_ *assets.StoreBuilder) (yaml.MapItem, error) {
```
Comment (ArthurSens) [inline pkg/prometheus/promcfg.go:4149]: If the Storebuilder isn't used, could we remove it from the arguments' list?

**Review by ArthurSens — COMMENTED** [review]
Just one small comment, otherwise LGTM

File: pkg/prometheus/promcfg.go [hunk 9]
```diff
@@ -2557,7 +2566,20 @@ func (cg *ConfigGenerator) GenerateAgentConfiguration(
 
 	// Remote write config
 	if len(cpf.RemoteWrite) > 0 {
-		cfg = append(cfg, cg.generateRemoteWriteConfig(store))
+		otlpcfg, err := cg.generateOTLPConfig(store)
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2569]: there's a glitch here?

File: pkg/prometheus/promcfg.go (hunk 8 above)
Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:4149]: +1

File: pkg/prometheus/promcfg.go [hunk 10]
```diff
@@ -4124,6 +4146,27 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
 
+func (cg *ConfigGenerator) generateOTLPConfig(_ *assets.StoreBuilder) (yaml.MapItem, error) {
+	if cg.version.LT(semver.MustParse("2.54.0")) {
+		return yaml.MapItem{}, fmt.Errorf("OTLP configuration is only supported from Prometheus version 2.54.0")
+	}
+
+	cfg := yaml.MapSlice{}
+	otlpConfig := cg.prom.GetCommonPrometheusFields().OTLP
+
+	if len(otlpConfig.PromoteResourceAttributes) > 0 {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:4157]: we don't really need to check for length.

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 11]
```diff
@@ -2011,3 +2017,15 @@ type ScrapeClass struct {
...
+// OTLPConfig is the configuration for writing to the OTLP endpoint.
+//
+// +k8s:openapi-gen=true
+type OTLPConfig struct {
+	// List of OpenTelemetry Attributes that should be promoted to metric labels, defaults to none.
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +listType=set
+	// +optional
+	PromoteResourceAttributes []string `json:"promoteResourceAttributes,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:2030]: We need a length validation on each item because Prometheus forbids empty attributes.

File: pkg/prometheus/promcfg.go (hunk 8 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go:4149]: Thanks. Got it. Done.

File: pkg/prometheus/promcfg.go (hunk 9 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go:2569]: My bad....  Revert it done

File: pkg/prometheus/promcfg.go (hunk 10 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go:4157]: Thanks. Got it. Done.

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 11 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/prometheus_types.go:2030]: Thanks. Got it .
Add `+kubebuilder:validation:Items:MinLength=1`  validation

**Review by ArthurSens — APPROVED** [review]
LGTM!

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 11 above)
Comment (simonpasquier) [reply pkg/apis/monitoring/v1/prometheus_types.go:2031]: Thanks, today I learn about `+kubebuilder:validation:items:...` :)

**Review by simonpasquier — APPROVED** [review]
